### ISSUE TITLE
Maintenance of https://github.com/qwerty12/AutoHotkeyScripts

### ISSUE THREAD

**Issue by dumblob** [dumblob/tmp#4]
Hi @qwerty12 , I'd like to ask if there is any chance of getting https://www.autohotkey.com/boards/viewtopic.php?t=27709 incorporated into https://github.com/qwerty12/AutoHotkeyScripts ? I think it'd be great addition and would help some people to make it easier to reuse.

Thanks!

**Comment by qwerty12** [comment]
Hi,

I kept LogonDesktop stuff in [my gists](https://gist.github.com/qwerty12) because it was easier to embed on the forum. In any case, I have no desire to add anything to an archived repo. For that matter, I've never touched LogonDesktop.ahk since. I also don't understand what use anybody would have for it, TBH - the code's a mess. I do find myself going back to look at it, however, for WinAPI function definitions to copy...

**Comment by dumblob** [comment]
Perfect, thanks for the gist links!

> I also don't understand what use anybody would have for it,

I need a seamless way to use special input devices without having the "credentials provider tile/icon" under the name/password input field (otherwise it's annoying if not impossible to log in especially in an enterprise context when there is e.g. a second factor like smartcard without which it's impossible to log in).

Specifically imagine something for disabled and/or stenography - in which case the device communicates using some non-standard protocol and on the Windows side an app/daemon manages to convert it to a virtual HID which AKH then shall read from.

But there might be better ways I don't know of - I'm at the beginning of exploration. Feel free to share your experience :wink:.

**Comment by dumblob** [comment]
Wow, your gists (not just those for AHK) are pretty interesting! I wonder what you think about https://github.com/ililim/dual-key-remap/discussions/36#discussioncomment-1305480 - i.e. the RShift/LShift -> Shift mapping hardcoded in low-level layout handling code on Windows.

**Comment by qwerty12** [comment]
> I need a seamless way to use special input devices without having the "credentials provider tile/icon" under the name/password input field (otherwise it's annoying if not impossible to log in especially in an enterprise context when there is e.g. a second factor like smartcard without which it's impossible to log in).

Using AutoHotkey for that would be anything but seamless... (And you'd really need to check if passwords can be zeroed-out from its memory) Credential Providers are the supported and documented way for this; it's easy to get Windows to attempt an interactive login from the details provided programmatically. How would AHK, an external program, manage it? `SendInput`, MSAA/UIA to write the provided details into the logon fields? Would be more error prone. 

> I wonder what you think about ililim/dual-key-remap#36 (comment) - i.e. the RShift/LShift -> Shift mapping hardcoded in low-level layout handling code on Windows.

I was never enough into `vim` to find use of such a utility but respect for writing it in a compiled language :-) The keyboard hook is one of AutoHotkey's operation modes.

**Comment by dumblob** [comment]
> Using AutoHotkey for that would be anything but seamless... (And you'd really need to check if passwords can be zeroed-out from its memory) Credential Providers are the supported and documented way for this; it's easy to get Windows to attempt an interactive login from the details provided programmatically.

I'm not enough experienced with Credential Providers, but my understanding is that a person has to first explicitly choose which provider to use from the list of icons under the input form. And this is the problem I see - ordinary users have absolutely no clue which provider to choose and why (i.e. "according to which criteria"). And this difficulty only because there is a convenient keyboard layout daemon they use.

>How would AHK, an external program, manage it? `SendInput`, MSAA/UIA to write the provided details into the logon fields? Would be more error prone.

Yep. But wouldn't need to confuse users with some weird Credential Providers icons and related stuff. Tough choice...

> I was never enough into `vim` to find use of such a utility but respect for writing it in a compiled language :-) The keyboard hook is one of AutoHotkey's operation modes.

:laughing: that utility has nothing to do with `vim` nor any other editor, but I get your point :wink:

I meant my question rather to find out how would you approach coping with the impossibility to distinguish RShift from LShift in a "modifier mode".

Btw. I just came up with yet another potential hack - namely to use custom written IME for this (invisible, doing just the RShift- & RShift- specific mappings and no more; should be just one tiny DLL) and leaving the rest to some standard keyboard layout. I've never worked with IME, so I can't tell whether it's as seamless as it sounds nor whether this'll work painlessly "by default" on a login screen.

**Comment by qwerty12** [comment]
>I'm not enough experienced with Credential Providers, but my understanding is that a person has to first explicitly choose which provider to use from the list of icons under the input form. And this is the problem I see - ordinary users have absolutely no clue which provider to choose and why (i.e. "according to which criteria").

I wrote one (well, modified Microsoft's existing sample) for a place and it's true some users don't know exactly what to click, but I reduced that happening somewhat by hiding users (actually the Credential Provider logged into an existing account). Windows 10 automatically selecting the last used "account" helps too.

>Yep. But wouldn't need to confuse users with some weird Credential Providers icons and related stuff. Tough choice...

You'll probably appreciate going with a documented solution when it comes time for a Windows upgrade. I would never rely on sending clicks/tabs etc. when a documented solution exists. Also, from what I remember, Windows 8 and 10's logon screen runs on the highest Window Band; any windows displayed by your script will only be shown once the user Alt+Tabs and activates them manually. Look at malcev's posts on the AHK forum - he may or may not have worked out how to create a window with a higher Band in AHK.

>that utility has nothing to do with vim nor any other editor, but I get your point

I thought the only reason people bothered remapping Caps Lock was 'cause of vim, TBH... Never understood it myself

>I meant my question rather to find out how would you approach coping with the impossibility to distinguish RShift from LShift in a "modifier mode".

I wouldn't know how to. ;) That's the beauty of AHK: robust keyboard hooking code that takes care of side effects and with its own workarounds where needed.

**Comment by dumblob** [comment]
>I wrote one (well, modified Microsoft's existing sample) for a place and it's true some users don't know exactly what to click, but I reduced that happening somewhat by [hiding users](https://www.windowscentral.com/how-hide-specific-user-accounts-sign-screen-windows-10) (actually the Credential Provider logged into an existing account). Windows 10 automatically selecting the last used "account" helps too.

Thanks for the valuable tip. I understand that the autoselection of the last used "account" includes also selection of the Credential Provider. That'd actually be enough I'd say. So it seems CP is a viable way forward (though maybe more work than IME or maybe not).

>You'll probably appreciate going with a documented solution when it comes time for a Windows upgrade. I would never rely on sending clicks/tabs etc. when a documented solution exists. Also, from what I remember, Windows 8 and 10's logon screen runs on the highest Window Band; any windows displayed by your script will only be shown once the user Alt+Tabs and activates them manually. Look at malcev's posts on the AHK forum - he may or may not have worked out how to create a window with a higher Band in AHK.

Yep, that could defeat the purpose indeed. Thanks for pointing it out as it seems the future direction of Windows is not to support such additions.

**Comment by qwerty12** [comment]
> I understand that the autoselection of the last used "account" includes also selection of the Credential Provider. That'd actually be enough I'd say.

Yes, all the user has to do is click the sign in button and enter the details as requested. 

> So it seems CP is a viable way forward (though maybe more work than IME or maybe not).

It's honestly been years since I looked at it, but the sample is very well documented - I think most people base their own Providers off it, as it beats redefining all the COM interfaces over again. The example was in C++, which is what I used, but it it is possible to write one in C# if that's your thing.

> Thanks for pointing it out as it seems the future direction of Windows is not to support such additions.

I think Credential Providers were their attempt to standardise things. Indeed, I wrote mine for Windows 7 and it worked on Windows 10 seamlessly. (XP's GINA, the predecessor, was something out of the wild wild west though - I have memories of using a replacement logon screen for XP implemented in Shockwave Flash!)
